namui-prebuilt/src/list_view.rs

```diff
@@ -0,0 +1,113 @@
+use crate::scroll_view;
+use namui::prelude::*;
+
+/// ListView is a vertical list view with fixed height items.
+#[derive(Debug)]
+pub struct ListView {
+    scroll_view: scroll_view::ScrollView,
+}
+
+pub struct Props<TItem, TIterator, TItems, TItemRender>
+where
+    TIterator: ExactSizeIterator<Item = TItem>,
+    TItems: IntoIterator<Item = TItem, IntoIter = TIterator>,
+    TItemRender: Fn(TItem, usize) -> RenderingTree,
+{
+    pub x: f32,
+    pub y: f32,
+    pub height: f32,
+    pub scroll_bar_width: f32,
+    pub item_wh: Wh<f32>,
+    pub items: TItems,
+    pub item_render: TItemRender,
+}
+
+pub enum Event {}
+
+impl ListView {
+    pub fn new() -> Self {
+        Self {
+            scroll_view: scroll_view::ScrollView::new(),
+        }
+    }
+    pub fn update(&mut self, event: &dyn std::any::Any) {
+        self.scroll_view.update(event);
+    }
+    pub fn render<TItem, TIterator, TItems, TItemRender>(
+        &self,
+        props: Props<TItem, TIterator, TItems, TItemRender>,
+    ) -> RenderingTree
+    where
+        TIterator: ExactSizeIterator<Item = TItem>,
+        TItems: IntoIterator<Item = TItem, IntoIter = TIterator>,
+        TItemRender: Fn(TItem, usize) -> RenderingTree,
+    {
+        let items_iter = props.items.into_iter();
+        let item_len = items_iter.len();
+
+        if item_len == 0 {
+            return RenderingTree::Empty;
+        }
+
+        let scroll_y = self.scroll_view.scroll_y;
+        let visible_item_start_index = (scroll_y / props.item_wh.height).floor() as usize;
+        let visible_item_end_index =
+            ((scroll_y + props.height) / props.item_wh.height).ceil() as usize;
+        let visible_item_count = visible_item_end_index - visible_item_start_index + 1;
+
+        let visible_items = items_iter
+            .enumerate()
+            .skip(visible_item_start_index)
+            .take(visible_item_count);
+
+        let rendered_items = visible_items.map(|(index, item)| {
```

Comment:
Now it provides index of just visible items, not of all items.
Do we need index of visible items? Can you give me an example of when to use the index?

Reply:
Actually I'm not sure about it.
You know, in javascript, they give us index in `forEach` and `map`.

In rust, maybe they can use enumerate to get index of item if they really want.

```
items: items.iter().enumerate(),
```
it works with current code. hmm. Ok. I will change it as you suggest!